src/ring_pedersen.rs

```diff
@@ -0,0 +1,318 @@
+use crate::paillier::DecryptionKey;
+use crate::{
+    errors::Result,
+    utils::{modpow, random_plusminus_scaled, random_positive_bn},
+    zkp::{
+        piprm::{PiPrmInput, PiPrmProof, PiPrmSecret},
+        Proof,
+    },
+};
+use libpaillier::unknown_order::BigNumber;
+use rand::{CryptoRng, RngCore};
+use serde::{Deserialize, Serialize};
+
+/// A commitment scheme based on a ring-variant of the Pedersen commitment scheme.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct RingPedersen {
+    /// The RSA modulus, corresponding to `N` in the paper.
+    modulus: BigNumber,
+    /// Ring-Petersen parameter `s`.
+    s: BigNumber,
+    /// Ring-Petersen parameter `t`.
+    t: BigNumber,
+}
+
+/// A [`RingPedersen`] commitment scheme alongside a zero knowledge proof that
+/// the commitment scheme was constructed correctly.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub(crate) struct VerifiedRingPedersen {
+    /// The underlying commitment scheme.
+    scheme: RingPedersen,
+    /// The zero knowledge proof that validates the correctness of [`VerifiedRingPedersen::scheme`].
+    proof: PiPrmProof,
+}
+
+/// A commitment produced by [`RingPedersen::commit`].
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct Commitment(BigNumber);
+
+impl Commitment {
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes()
+    }
+}
+
+/// The randomness generated as part of [`RingPedersen::commit`].
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[repr(transparent)]
+pub(crate) struct CommitmentRandomness(BigNumber);
+
+impl CommitmentRandomness {
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes()
+    }
+
+    /// Masks randomness with `mask` and `challenge`.
+    /// The output [`MaskedRandomness`] value is computed as `mask - challenge * value`.
+    pub(crate) fn mask_neg(
+        &self,
+        mask: &CommitmentRandomness,
+        challenge: &BigNumber,
+    ) -> MaskedRandomness {
+        MaskedRandomness(&mask.0 - challenge * &self.0)
+    }
+
+    /// Masks randomness with `mask` and `challenge`.
+    /// The output [`MaskedRandomness`] value is computed as `mask + challenge * value`.
+    pub(crate) fn mask(
+        &self,
+        mask: &CommitmentRandomness,
+        challenge: &BigNumber,
+    ) -> MaskedRandomness {
+        MaskedRandomness(&mask.0 + challenge * &self.0)
+    }
+
+    /// Returns the randomness as a [`MaskedRandomness`].
+    pub(crate) fn as_masked(&self) -> &MaskedRandomness {
+        // Safety: This is safe because both CommitmentRandomness and MaskedRandomness
+        // are `repr(transparent)` and contain the same underlying data type.
+        unsafe { std::mem::transmute(self) }
```

Comment:
People tend to get scared by `unsafe` code (probably rightfully so) so if we can avoid it all the better. `bytemuck` is a fairly standard library and would allow us to avoid this, so that might be the best approach.